### PR TITLE
Extend Curve to allow for domains outside of `[0, 1]`.

### LINKED ISSUE #5607: Extend Curve and CurveEditor to allow for a domain beyond [0, 1]
### Describe the project you are working on

2D semi-realistic flight simulator using coefficient of lift curves to define wing lift.

### Describe the problem or limitation you are having in your project

Coefficients of lift are a function of angles of attack (AOA), which range over [-180, 180]. In order to use the CurveEditor to define these coefficient of lift curves for different airfoils/airplanes, I would have to manually guesstimate where in [0, 1] a specific AOA like -10 or 6 would be before adding a point in the CurveEditor.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

This is a proposal to extend the hardcoded domain of Curve to user-defined limits. This is already possible for its range, but not its domain.

This allows anyone to define curves over the domain that makes sense for their own application. Coefficients of lift based off AOA, acceleration based off current speed, a spring's pushback depending on current compression, jump height over time, etc etc.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

We could modify and extend the Curve class to generalize it to arbitrary `real_t` domains. We would expose those `min_x` and `max_x` properties to users through the inspector, like `min_y` and `max_y` already are.

Most of what needs to happen everywhere for this to work is already implemented in the Y coordinates. We just need to implement it for the X coordinates.

I have noticed that Curve is used in a few other places (e.g. animations, particles), so there is an alternative that leaves `Curve` alone so as not to break compatibility or cause unwanted issues/confusion. We could e.g. rename `Curve` to `UnitCurve` and subclass it with a new `Curve` class containing the extra methods; or subclass `Curve` into `ArbitraryCurve` or some other suitable name.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

It is possible to write scripts that convert values from an arbitrary domain into the unit domain [0,1].

However, this wouldn't allow someone to directly modify a Curve (its points and tangents), through the CurveEditor, within the domain that is relevant to their application.

### Is there a reason why this should be core and not an add-on in the asset library?

We would be extending the functionality of something already existent in core, so it doesn't make sense to reimplement the same (slightly extended) feature in another language and have it be an option in alternative to something already in core.

### PR DESCRIPTION
**Pull request by anvilfolk**
This PR allows users to make curves in whatever app/game-specific domain they wish 📈 Previously, curves were restricted to a domain of [0,1]. Implements proposal, closes [#5607](https://github.com/godotengine/godot-proposals/issues/5607).

Depends on #78931 and #77622, whose commits have been cherry-picked in this PR. Can rebase/remove those commits when/if they get merged. I can also squash them into this PR if preferred. Let me know 🥰 

https://github.com/godotengine/godot/assets/1133892/a0e957fc-b914-4e44-8b4f-b3bb27687ca0

⚠️ This PR also enforces previously unenforced value ranges for points in the curve ⚠️ I tested this and it should _not_ alter existing curve's points when they are loaded ⚠️ Instead, this happens, then is corrected when limits are changed:

image

### Things accounted for

- Curve editor supports extended curves
- Preview generates correctly
- Baking works correctly
- Presets are set correctly in new domains
- Duplicated unit curve test cases adapted to a domain of [-100, 100]
- Documented that other uses of Curve in the engine require unit curves

### CODE REVIEW

**Comment by anvilfolk** [comment]
Compilation is failing due to compiler not knowing whether to use lerp(floats) or lerp(doubles). I don't see that warning or error locally. I'm compiling on Visual Studio 2022 with -dev_mode, but that's about it. Is there any flag I ought to be using?

I think it also speaks as to whether we should make a decision to use real_t, float or double for these values. Or maybe just not use lerp at all?

**Comment by anvilfolk** [comment]
Fixed all the issues that were showing up in the checks! I'd love to work on some UI/UX work on the editor itself next , but that only makes sense once this PR is mostly settled :) Looking at https://github.com/godotengine/godot-proposals/issues/5579 and https://github.com/godotengine/godot-proposals/issues/3267, for example!

So LMK if there's any issues or feedback <3

File: scene/resources/curve.h
Comment (Iniquitatis) [inline scene/resources/curve.h:102]:
```suggestion
	real_t get_value_range() const { return _max_value - _min_value; }
```

Comment (anvilfolk) [reply scene/resources/curve.h]: Thank you, great catch!!! <3

File: editor/plugins/curve_editor_plugin.cpp
Comment (MewPurPur) [inline editor/plugins/curve_editor_plugin.cpp:444]: Woop, I gotta assume this is meant to be "Add Curve Point"?

**Comment by anvilfolk** [comment]
Converted to draft since I'm fairly certain this isn't the best way to do it. Will work on it once 4.0 is released!

**Comment by akien-mga** [comment]
Is this superseded by #74959?

**Comment by anvilfolk** [comment]
It is not! :) It still very much is something I want to do, but I believe I have a better approach and I will work on reimplementing it as time allows. Might be a while still.

**Comment by anvilfolk** [comment]
Finally fixed and reopened this PR! I updated the original post with information.

Couple of changes:
- Opted out of having a `unit_curve` parameter. This means it's less clutter in the inspector.
- It's been updated to @MewPurPur 's revamp of the curve editor in #74959
- Depends on a couple of other PRs that facilitate this one, but can be squashed into this one if you want, though they are all relatively different issues.

Most of the lines of code are documentation and tests. The changes themselves are relatively small, especially once you remove the other two cherry-picked commits.

I have ⚠️ not tested anything other than the curve editor in the inspector ⚠️ , though apparently curves are used in other places. The documentations do reflect the idea that most curves used in the engine need to be unit curves so 🤷 😅 ?

I believe CI is passing, I just seem to have a perpetual single build type that fails?

**Review by MewPurPur — CHANGES_REQUESTED** [review]
The code that exists LGTM, however, additional changes are needed to make domains play nicely with existing functionality.

The hidden Alt functionality still considers that 0 and 1 are the edges, and so if you use it on a leftmost or rightmost point, it will be forced into that part of the domain.

Snapping doesn't work properly, because it still snaps on the x-axis based on `1.0/snap_divisions.x`. When you fix this, remember that you need to do it in two places, once for the dragging logic, and once for the adding a point logic.

Shift for snapping to an axis will also draw the horizontal line only from 0 to 1.

File: editor/plugins/curve_editor_plugin.h
Comment (anvilfolk) [reply editor/plugins/curve_editor_plugin.h:143]: This was just to silence uninitialized class variable warnings! I added what seemed to be some sensible default values, but I don't think I want to change the names :) Does that make sense?

File: scene/resources/curve.cpp
Comment (anvilfolk) [reply scene/resources/curve.cpp:696]: Because we don't really know what people's domain is going to be, and I didn't want to limit people. When I played around with it, the hint range was fairly stringent. I didn't immediately find a way to go beyond the hint range. So despite it making the UX not great because click-dragging makes huge values, that still felt better than it being hard/impossible to go beyond a randomly chosen limit.

**Comment by anvilfolk** [comment]
> The code that exists LGTM, however, additional changes are needed to make domains play nicely with existing functionality.
> 
> The hidden Alt functionality still considers that 0 and 1 are the edges, and so if you use it on a leftmost or rightmost point, it will be forced into that part of the domain.
> 
> Snapping doesn't work properly, because it still snaps on the x-axis based on `1.0/snap_divisions.x`. When you fix this, remember that you need to do it in two places, once for the dragging logic, and once for the adding a point logic.
> 
> Shift for snapping to an axis will also draw the horizontal line only from 0 to 1.

Thank you for finding this! I played around a little with it and it seemed to be working, but I clearly missed something! I'll take a look when I get time! :)

File: editor/plugins/curve_editor_plugin.h
Comment (MewPurPur) [reply editor/plugins/curve_editor_plugin.h:143]: Sure! The defaults look good.

File: scene/resources/curve.cpp
Comment (MewPurPur) [reply scene/resources/curve.cpp:696]: Use the "or_greater" and "or_lesser" hints; hit me up if you don't figure it out

Comment (anvilfolk) [reply scene/resources/curve.cpp]: Did it! :)

**Comment by anvilfolk** [comment]
Last push should have fixed:

- Snapping not working properly (dragging and point creation)
- Snapping axes showing improperly
- Alt-drag not keeping point between its two original neighbors

**Review by MewPurPur — APPROVED** [review]
The curve editor changes are good. The code in general also looks good to me.

**Comment by mbrannen** [comment]
bumping because I'd like this feature to be included

**Comment by AThousandShips** [comment]
Please don't bump PR without contributing significant new information. Use the :+1: reaction button on the first post instead.

File: doc/classes/Curve.xml
Comment (ttencate) [inline doc/classes/Curve.xml:167]: Maybe add `(y-coordinate)` after the word `value`. Same below.

File: scene/resources/curve.cpp
Comment (ttencate) [inline scene/resources/curve.cpp:696]: It might make sense to move domain above value, so X comes before Y in the editor.

**Review by ttencate — COMMENTED** [review]
I have wanted this feature for months, was happy to discover this proposal and PR and would _very_ much like it to land in 4.3, and I'm willing to help make that happen. So here's an unsolicited review.

For what it's worth, the code changes look good to me. The documentation changes are especially great. Succinct, yet clear and unambiguous.

The merge conflicts are simple to resolve:

- replace `mm->is_ctrl_pressed()` by `mm->is_command_or_control_pressed()` in two places
- replace `SNAME("Editor")` by `EditorStringName(Editor)` in two places

I tested it manually in the editor and didn't find any issues apart from rendering and editing the curve if max domain or max value is set extremely large (1e10). But I don't think that should be a blocker.

**Comment by anvilfolk** [comment]
Thanks for taking a look @ttencate!

I've been wanting to come back to this. Right now I believe the main blocker is that it depends on #78931, which @YuriSizov believes should be handled differently.

I'm starting to look into how we can add another `ADD_PROPERTY_*` macro which allows for setters to be used only during resource loading, but I'm not sure how long it'll take me to figure this one out :)

I do still think it'd be OK to merge as-is, since the compatibility breaking is very minor and should not affect saved settings even after (and other resources are using the same trick), but I understand it'd be wonderful to have a proper solution :)

**Comment by ttencate** [comment]
Does this PR actually need to depend on #78931? Or is that just to avoid merge conflicts down the line?

**Comment by anvilfolk** [comment]
@ttencate it does depend on it because while the buggy loading of min/max values isn't a big deal when they're not actually enforced, they are and need to be enforced for curve domains. So if the domain numbers load wrongly, the curve will unusable :(

**Comment by ttencate** [comment]
One way to unblock this is to leave the bug be, and print a runtime error if a curve with an invalid domain is used (e.g. sampled or rendered in the curve editor). The erroneous state only happens when the user sets the domain right after creating a new curve. At that moment, they'll probably enter both min and max right away, and if they don't, they'll notice soon enough that the curve cannot be edited. I don't think that's a big deal. We could even have the curve visualization show something like "Invalid domain".

The advantage is that this leaves the path free for a cleaner solution later on, unlike #78931 which isn't backwards compatible.

**Comment by ttencate** [comment]
@anvilfolk What do you think of [the above idea](https://github.com/godotengine/godot/pull/67857#issuecomment-1895414704)? I'm willing to implement it and create a PR against your branch if you don't have time for it. I really want this to land in 4.3 :slightly_smiling_face:

**Comment by ttencate** [comment]
> Having Godot load your perfectly valid curve with invalid domains that you need to fix manually every time you open it feels, to me, like a no-go.

That's not what I was suggesting. An invalid domain is saved and loaded as-is. You would only get a curve with an invalid domain upon loading if it was saved like that.

**Review by Mickeon — COMMENTED** [review]
Nothing against the feature at all, but the fact that this note has to be specified into every property that requires this "unit" Curve peeves me a lot.
Calling it "unit" Curve may not be incorrect, but it is a bit "meaningless" as it won't matter much to most readers. It won't matter nearly as much as something like a warning or anything more noticeable and self-explanatory.
Repetition is not just harder to translate but harder to maintain as well.

Worth noting, a similar issue occurred in https://github.com/godotengine/godot/pull/85207 . There it was concluded to generate the note automatically, but that may not be as easy here.

**Comment by anvilfolk** [comment]
Yeah, it is definitely annoying :( Do you have any specific suggestions? The entire engine kind of expects a unit curve because that's what was available so far, and behavior is undefined when non-unit curves are passed to these things.

I don't think we can automate the comments easily, since both curves and unit curves can now be used in different parts of Godot.

I see two possibilities: 1) Add a `is_unit_curve` boolean to `Curve` that enforces the domain or, 2) Create a `UnitCurve` subclass of `Curve` that enforces the domain. However, both would still require us to change code across the entire codebase to ensure the components that expect unit curves are actually getting unit curves.

**Review by Mickeon — COMMENTED** [review]
Some things before getting to the nitty-gritty that are the docs right now because that's my biggest peeve right now.

I think most of what's necessary is specifying "unit [Curve]", without ruining the flow of the sentence.
I feel like changing the "domain" for all of those properties is fairly unlikely and potentially worth noting in the **Curve** docs themselves. Plus, actually doing it should warrant a warning _of some kind_ (WAY outside the scope of this PR).

**Comment by anvilfolk** [comment]
Rebased and implemented @Mickeon 's suggestions :)

**Review by fire — APPROVED** [review]
I support having this feature, but I cannot do a complete technical review now.

File: scene/resources/curve.cpp
Comment (AThousandShips) [inline scene/resources/curve.cpp:695]:
```suggestion
	ADD_PROPERTY(PropertyInfo(Variant::FLOAT, "min_domain", PROPERTY_HINT_RANGE, "-1024,1024,0.01,or_greater,or_less", PROPERTY_USAGE_EDITOR), "set_min_domain", "get_min_domain");
	ADD_PROPERTY(PropertyInfo(Variant::FLOAT, "max_domain", PROPERTY_HINT_RANGE, "-1024,1024,0.01,or_greater,or_less", PROPERTY_USAGE_EDITOR), "set_max_domain", "get_max_domain");
	ADD_PROPERTY(PropertyInfo(Variant::FLOAT, "min_value", PROPERTY_HINT_RANGE, "-1024,1024,0.01,or_greater,or_less", PROPERTY_USAGE_EDITOR), "set_min_value", "get_min_value");
	ADD_PROPERTY(PropertyInfo(Variant::FLOAT, "max_value", PROPERTY_HINT_RANGE, "-1024,1024,0.01,or_greater,or_less", PROPERTY_USAGE_EDITOR), "set_max_value", "get_max_value");
```

**Review by AThousandShips — APPROVED** [review]

**Comment by anvilfolk** [comment]
Thanks for the reviews! Excited to see where extended curves may be used elsewhere in the engine :)

Let me know if you want me to squash the 3 commits (which are includes for the two other ancillary PRs).

**Comment by fire** [comment]
We prefer squashed commits. For future prs, if you want to separate them, use different pull requests.

**Comment by KoBeWi** [comment]
They are different pull requests already though.

**Comment by anvilfolk** [comment]
It was mostly to separate discussion & ease of reviewing. It sounds like reviewers are OK with them, so I'll go ahead and squash, especially since some of the latest changes are in the latest commit but really should've applied to other PRs.

**Comment by anvilfolk** [comment]
Looks like I messed up the commits, as usual. Will try fixing soon 🤦

**Review by Calinou — APPROVED** [review]
Tested locally, it works as expected. Code looks good to me.

One potential UX concern is that the concept of *unit curve* isn't enforced in the inspector, so many nodes will expose the domain properties within the Curve resources even though they shouldn't. This adds visual noise to the inspector, but also room for accidental edits.

The best way to resolve this would be to add a new property hint similar to `PROPERTY_HINT_COLOR_NO_ALPHA`, except that `PROPERTY_HINT_RESOURCE_TYPE` is already used here to specify Curve as a valid resource type.

Nonetheless, I'll still approve this PR considering it improves Curve's usability for game design use cases, particularly for users who are not writing the scripts that handle the curves in question.

**Testing project:** [test_pr_67857.zip](https://github.com/user-attachments/files/17837673/test_pr_67857.zip)

**Comment by anvilfolk** [comment]
Thank you for the review!

I only found unsatisfactory solutions to that problem, unfortunately, though I agree it exists.

We could make this into an `ArbitraryCurve` that `Curve` inherits from. However, `Curve` would then simply ignore/throw errors at attempted domain changes. So there's useless, confusing methods. It would should keep backwards compatibility this way though.

Having `ArbitraryCurve` inherit from the current `Curve` would mean that any place that requires a curve could be assigned an arbitrary one, so it doesn't solve the problem.

The other solution, which is perhaps most viable, is to add a "unit" toggle for curves, which enables domain properties. Every place that uses curves would still need to check & enforce that it is using unit curves though, so it only solves the problem partially.

**Comment by ttencate** [comment]
`ArbitraryDomainCurve` cannot inherit from `UnitCurve`, nor the other way round, because the Liskov substitution principle doesn't apply in either direction:

- `ArbitraryDomainCurve` cannot inherit from `UnitCurve`, because it would let you set an `ArbitraryDomainCurve` where only a `UnitCurve` is valid.
- `UnitCurve` cannot inherit from `ArbitraryDomainCurve`, because it doesn't let you change the domain.

So a correct inheritance hierarchy would be (naming aside):

```
- Resource
  `- AbstractCurve
     |- UnitCurve (what's currently called Curve)
     `- ArbitraryCurve
```

The shared functionality would be in `AbstractCurve`, on which `min_domain` and `max_domain` are abstract virtual methods. `UnitCurve` would override these to just return 0 and 1, whereas `ArbitraryCurve` would override them to read from fields, and also provide `set_min_domain` and `set_max_domain`.

If `UnitCurve` would just be called `Curve`, this change is _almost_ backwards compatible; the insertion of `AbstractCurve` into the inheritance chain does change reflection/introspection, and probably also has backwards compatibility problems in GDExtension.

I don't think this discussion warrants holding up the PR further, though. Just wanted to write it down for posterity.

**Comment by anvilfolk** [comment]
Oh, excellent point, hadn't thought of it at all, thanks for bringing it up.

I think it's worth taking time to do well though, because otherwise restructuring to something sensible probably only happens for Godot 5.

So the question is whether it's worth refactoring things in the way you describe. We could keep a unit curve named `Curve` for backwards compatibility. This should keep all current functionality

**Comment by anvilfolk** [comment]
Mistakenly closed because toddler shenanigans 😅

**Comment by Mickeon** [comment]
> The best way to resolve this would be to add a new property hint similar to `PROPERTY_HINT_COLOR_NO_ALPHA`, except that `PROPERTY_HINT_RESOURCE_TYPE` is already used here to specify Curve as a valid resource type.

Only tangentially related, but this falls into the idea I've been concocting for a really long time, involving turning `hint` and `hint_string` into multiple components (basically a Dictionary of hints). This is very far off in the future, though, so I wouldn't worry about it much. I feel like it's a problem that may solve itself in the future.
At worst, there could be a configuration warning for all affected properties.

**Comment by akien-mga** [comment]
Thanks! And congrats for ~your first merged~ **the merge of your first** Godot contribution :tada:

**Comment by anvilfolk** [comment]
Oh my goodness! This legitimately feels so nice :) Very excited to see all the regressions from arbitrary curves making their sneaky way around Godot, hehehe 🤣

**Comment by ttencate** [comment]
Just want to say a huge THANK YOU to @anvilfolk for seeing this through! I'm using this everywhere now and it makes my code and resources _so_ much simpler. Amazing work! 🤩

**Comment by anvilfolk** [comment]
Thank you for pushing for it too, that helped! This feature was why I originally started contributing to Godot, so it feels really nice to see it in ❤️

**Pull request merged**
